// parse-server/lib/GraphQL/loaders/schemaDirectives.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.load = exports.definitions = void 0;
var _utils = require("@graphql-tools/utils");
var _FunctionsRouter = require("../../Routers/FunctionsRouter");
const definitions = exports.definitions = `
  directive @resolve(to: String) on FIELD_DEFINITION
  directive @mock(with: Any!) on FIELD_DEFINITION
`;
const load = parseGraphQLSchema => {
  parseGraphQLSchema.graphQLSchemaDirectivesDefinitions = definitions;
  const resolveDirective = schema => (0, _utils.mapSchema)(schema, {
    [_utils.MapperKind.OBJECT_FIELD]: fieldConfig => {
      var _getDirective;
      const directive = (_getDirective = (0, _utils.getDirective)(schema, fieldConfig, 'resolve')) === null || _getDirective === void 0 ? void 0 : _getDirective[0];
      if (directive) {
        const {
          to: targetCloudFunction
        } = directive;
        fieldConfig.resolve = async (_source, args, context, gqlInfo) => {
          try {
            const {
              config,
              auth,
              info
            } = context;
            const functionName = targetCloudFunction || gqlInfo.fieldName;
            return (await _FunctionsRouter.FunctionsRouter.handleCloudFunction({
              params: {
                functionName
              },
              config,
              auth,
              info,
              body: args
            })).response.result;
          } catch (e) {
            parseGraphQLSchema.handleError(e);
          }
        };
      }
      return fieldConfig;
    }
  });
  const mockDirective = schema => (0, _utils.mapSchema)(schema, {
    [_utils.MapperKind.OBJECT_FIELD]: fieldConfig => {
      var _getDirective2;
      const directive = (_getDirective2 = (0, _utils.getDirective)(schema, fieldConfig, 'mock')) === null || _getDirective2 === void 0 ? void 0 : _getDirective2[0];
      if (directive) {
        const {
          with: mockValue
        } = directive;
        fieldConfig.resolve = async () => mockValue;
      }
      return fieldConfig;
    }
  });
  parseGraphQLSchema.graphQLSchemaDirectives = schema => mockDirective(resolveDirective(schema));
};
exports.load = load;
